cayenne/src/main/java/org/apache/cayenne/access/flush/RootRowOpProcessor.java

```diff
@@ -74,14 +79,37 @@ public Void visitUpdate(UpdateDbRowOp dbRow) {
 
     @Override
     public Void visitDelete(DeleteDbRowOp dbRow) {
-        if (dbRowOpFactory.getDescriptor().getEntity().isReadOnly()) {
+        ObjEntity entity = dbRowOpFactory.getDescriptor().getEntity();
+        if (entity.isReadOnly()) {
             throw new CayenneRuntimeException("Attempt to modify object(s) mapped to a read-only entity: '%s'. " +
-                    "Can't commit changes.", dbRowOpFactory.getDescriptor().getEntity().getName());
+                    "Can't commit changes.", entity.getName());
         }
         diff.apply(deleteHandler);
-        Collection<ObjectId> flattenedIds = dbRowOpFactory.getStore().getFlattenedIds(dbRow.getChangeId());
-        flattenedIds.forEach(id -> dbRowOpFactory.getOrCreate(dbRowOpFactory.getDbEntity(id), id, DbRowOpType.DELETE));
-        if (dbRowOpFactory.getDescriptor().getEntity().getDeclaredLockType() == ObjEntity.LOCK_TYPE_OPTIMISTIC) {
+
+        DbEntity dbSource = entity.getDbEntity();
+        Set<Entry<CayennePath,ObjectId>> flattenedEntries = dbRowOpFactory.getStore().getFlattenedEntries(dbRow.getChangeId());
+
+        flattenedEntries.forEach(entry -> {
+            DbRelationship dbRel = dbSource.getRelationship(entry.getKey().first().toString());
+
+            // Don't delete if the target entity has a toMany relationship with the source entity,
+            // as there may be other records in the source entity with references to it.
+            if (!dbRel.getReverseRelationship().isToMany()) {
+
+                // Check if there's an ObjRelationship matching the flattened attributes DbRelationship
+                boolean hasObjRelationship = entity.getRelationships().stream()
```

Comment:
This looks heavy, though not many flattened attributes are usually mapped. Ultimately this could be precalculated and moved to the model (`ClassDescriptor.additionalDbEntities` maybe?), but that is definitely 5.0 task only, while this fix could be easily backported to 4.2

Reply:
I'd prefer this being backported to 4.2 if possible.

Reply:
Sure, let's keep it as is for 4.2 compatibility.